### DISCUSSION TITLE
Typescript error following docs example

### DISCUSSION THREAD

**Discussion by ShlomoNus** [pmndrs/jotai#1870] [Q&A]
```ts
const strAtom = atom('') // primitive string atom

const readWriteAtom = atom<string, number>(
  (get) => get(strAtom),
  (_get, set, num) => set(strAtom, String(num))
)
```
I get an error:
No overload expects 2 type arguments, but overloads do exist that expect either 1 or 3 type arguments.ts(2743)

I can fix it by omitting <string, number> but what the point?

**Comment by dai-shi** [comment]
If you want explicitly type such writable atoms, you can't omit types. The third argument should be the return type of `write` function.
```ts
const strAtom = atom('') // primitive string atom

const readWriteAtom = atom<string, number, void>(
  (get) => get(strAtom),
  (_get, set, num) => set(strAtom, String(num))
)
```

**Reply by ShlomoNus** [reply]
Oh got you, but I got not this error :
Type 'number' does not satisfy the constraint 'unknown[]'.ts(2344), so set the atom generic type to :atom<string, number[], void> and now it work as charm.
Can you the of official [docs](https://jotai.org/docs/guides/typescript#derived-atoms-are-also-type-inferred-and-explicitly-typed)
as this example from the there.

**Reply by dai-shi** [reply]
Nice catch! Would you like to open a PR please?
We should recommend inferring types:
```ts
const asyncStrAtom = atom(async () => 'foo')
const writeOnlyAtom = atom(null, (_get, set, str: string) => set(fooAtom, str))
const readWriteAtom = atom(
  (get) => get(strAtom),
  (_get, set, num: number) => set(strAtom, String(num))
)
```

**Reply by eikonomega** [reply]
@dai-shi I've created a pull request which I think will make this much clearer in the documentation.
It's currently waiting for review.

#1885
